wgpu/backends/wgpu_native/_helpers.py

```diff
@@ -234,49 +233,62 @@ class WgpuAwaitable:
     truely async manner, as well as to support a synchronous version of them.
     """
 
-    def __init__(self, title, result, callback, poll_function, finalizer, timeout=5.0):
+    def __init__(self, title, callback, finalizer, poll_function=None, timeout=5.0):
         self.title = title  # for context in error messages
-        self.result = result  # a dict that the callbacks writes to
         self.callback = callback  # only used to prevent it from being gc'd
-        self.poll_function = poll_function  # call this to poll wgpu
         self.finalizer = finalizer  # function to finish the result
+        self.poll_function = poll_function  # call this to poll wgpu
         self.maxtime = time.perf_counter() + float(timeout)
+        self.result = None
+
+    def set_result(self, result):
+        self.result = (result, None)
 
-    def is_done(self):
+    def set_error(self, error):
+        self.result = (None, error)
+
+    def _is_done(self):
         self.poll_function()
-        if self.result:
-            return True
-        if time.perf_counter() > self.maxtime:
-            self.result["timeout"] = True
-            return True
-        return False
-
-    def finish(self):
-        if "result" in self.result:
-            return self.finalizer(self.result["result"])
-        elif "error" in self.result:
-            raise RuntimeError(self.result["error"])
-        elif "timeout" in self.result:
-            raise RuntimeError(f"Waiting for {self.title} timed out.")
-        else:
-            raise RuntimeError(f"Failed to obtain result for {self.title}.")
+        return self.result is not None or time.perf_counter() > self.maxtime
+
+    def _finish(self):
+        try:
+            if not self.result:
+                raise RuntimeError(f"Waiting for {self.title} timed out.")
+            result, error = self.result
+            if error:
+                raise RuntimeError(error)
+            else:
+                return self.finalizer(result)
+        finally:
+            # Reset attrs to prevent potential memory leaks
+            self.callback = self.finalizer = self.poll_function = self.result = None
```

Comment:
FYI I added this because these functions are likely closures, so this is somewhat prone to holding references. I think this is the reason why I used that `result_dict` earlier.

Reply:
I've spent my life programming in Lisp, Java and Python rather than C, C++, Rust, etc.  I tend to (foolishly?) trust the GC to do its job better than I can.  Unless it's a big object, I tend not to worry.. The callback has a reference to the awaitable and the awaitable has a reference to the callback, but the GC will take care of that. 

However if that bothers you, I understand.

Reply:
The gc can handle circular refs, yeah, but they do take longer for the gc to remove. 

I also trust the gc in most code, but most code in wgpu is potentially called a lot of times per second, so I'm more on my guard as usual 😉